### DISCUSSION TITLE
Schmidt trigger for the next hardware rev?

### DISCUSSION THREAD

**Discussion by digiexchris** [kachurovskiy/nanoels#113] [Ideas]
I saw your historical repo of the ESP version, and I was also working on a TouchDRO at the same time. I noticed TouchDRO's esp implementation uses a Schmitt trigger for signal inputs to help clean things up and give a nice square edge, especially if the edge is noisy. I find it works really well with long wires with high capacitance where the pulses have a slower rise and get more sawtooth shaped, and the slight hysteresis helps a lot in noisy environments. it seems to be more esd resistant than the ESP32 seems to be also. Have you considered using one, such as the 17HC14? Could combine it with a simple transistor for level shifting too, to support encoder voltages up higher than 5v.

https://www.ti.com/lit/ml/scea046a/scea046a.pdf?ts=1680067926446

That brings up another idea: how about designing an adapter board that does a pin-translation to the nano, includes a schmitt trigger/level shifting circuit, and plugs straight in the same space on the H2. A simple define at the top of the pin configs could switch pins. Thoughts?

**Comment by kachurovskiy** [comment]
Hi Chris, I did see a slow rise of the encoder signals but I dealt with it in 2 ways:

1. Added 1kOhm pull-ups on the A and B lines, this helped a lot
2. Added a logic change where if the encoder pulses are coming fast enough, I stop reading the B line because spindle can't reverse the rotation direction due to the mechanical inertia - https://github.com/kachurovskiy/nanoels/blob/main/h2/h2.ino#L442
3. If you see high-frequency noise, inline ferrite bead is the usual way of dealing with it
4. For ESD protection, normally 6V ESD diodes to ground are used though NanoEls doesn't prescribe them in the [published schematics](https://github.com/kachurovskiy/nanoels/tree/main/h2) to keep things simple for those building it at home

If you think 17HC14 will benefit your project I have no objections though.

> That brings up another idea: how about designing an adapter board that does a pin-translation to the nano

What problem would that solve for the NanoEls H2? I don't think I saw reports saying they're losing encoder pulses or something like that. Or is this to use 1 encoder both with Nano and some other device e.g. TouchDRO? Thanks!

**Comment by digiexchris** [comment]
Oh that was purely to adapt an esp32 to the h2.

**Comment by digiexchris** [comment]
It could also allow additional periferals without needing to change the
main h2 board, such as physical end stop switches or output to a touchdro
or something like that, along with level shifters to make an esp
voltage-compatible.

**Reply by kachurovskiy** [reply]
Sorry, I might be missing the context you're coming from: what is that you're looking to build/achieve? Thanks.

**Comment by digiexchris** [comment]
I'm sorry I'm not very clear, I'm basically thinking out loud. 

I enjoy modularity and being able to piece gear together with whatever I have on hand, so I was thinking about 2 things at the same time and asked them in the same discussion. Firstly, I think schmitt triggers are the bees knees, but yes your solution for interferance works fine. Secondly, making the next version (h3?) somewhat modular so you can use adapter boards to add different MCUs or other filtering circuitry or whatever the user may feel the need for is a kind of design I like. 

For example, a header to plug in LCDs, but also a header for an I2C lcd as well. I have an I2C version of the 2004, 4 wires, really easy. Saves some pins if someone wants to use an esp8266 or something. 

I first started down the world of stepper control via the 3d printing world, and the early 3d printing firmwares tended to attempt to allow people to piece stuff together with whatever they could find, which is definitely not a problem here for this project because all of these components are readily available.

Just a friendly design discussion, no need to actually solve problems :) Feel free to close it.

I did whip up a home etched version of the H2 board, and I've gotta say, it works really well. There are a few tracks that are a little close together that didn't etch well, I had to manually repair those, but it wasn't a big deal. Have you considered selling parted out kits for people to solder together? I think they could sell well.

**Reply by kachurovskiy** [reply]
Got it, thank you :) Happy to hear that it worked. Coincidentally, I just started selling kits for H2 on kachurovskiy.com yesterday 😁

I love modularity as well but production costs and regulatory environment across the world doesn't favor it for electronics. Each board / variation has a high fixed cost. Fewer parts, automated manufacturing and less things to certify for EMC seems to be the way to profitability (I'm not there by a long shot 😑).

I used I2C adapter for the display before but:

- it takes longer to update the screen via I2C (52ms or something like that) than via 4 pins, 8 pins is 2x faster still
- adapter removes physical connection of the screen to the PCB which I'm currently relying on
- adapter is another component to place and take care of

H3 is already in the works. You can check out a short video about it on https://t.me/MaximKachurovskiy if you're interested. Cheers!

**Reply by digiexchris** [reply]
very nice!

I know it increases the cost, but how do you feel about keyboard switches for buttons? I like large buttons for machining type stuff where I'm watching the machine more than the UI.  
![image](https://user-images.githubusercontent.com/5871642/229306991-fdde08fd-52dc-443e-910b-fd939e468c2c.png)

**Reply by kachurovskiy** [reply]
Yeah, buttons are are topic that can be hotly debated on any day :) I like this style of buttons as well, in fact I'm typing on them right now - but I've never seen a CNC controller using something like this. I believe the reasons for that are:

- lack of dust / liquid protection
- explosion in number of parts due to key caps
- presence of hysteresis when it's half-clicked
- huge space requirements
- high cost

H2 graphics overlay I sell has none of the aforementioned issues while close in user experience:

- tactile feedback before pressing (embossing)
- tactile and sound feedback when pressed
- custom graphics

For some people the "click" mechanics of a keyboard switch is very important and I respect that but I don't see a good way to provide it.
